[auth] added

cseed opened the pull request:
testing

jigold commented on hail/python/hailtop/gear/deploy_config.py:
```diff
@@ -0,0 +1,57 @@
+    def domain(self, service, location=None):
+        if not location:
+            location = self.location
+        ns = self._service_ns(service)
```
should this be _service_namespace?

cseed replied:
Sorry, my names aren't great.  `default_namespace` is the namespace to expect a given service in by default.  `service_namespace` is a set of overrides for specific services.  I want to support that so I can e.g. deploy batch in my cseed, but use the auth service in default.  Make sense?  So everyone should go through _service_ns.

jigold commented on hail/python/hailtop/gear/deploy_config.py:
```diff
@@ -0,0 +1,57 @@
+        return 'internal.hail.is'
+
+    def base_path(self, service):
+        ns = self._service_ns(service)
```
Same about service_namespace vs service_ns

cseed commented:
This isn't a hard change, but it is a big one.  Let me know if you want me to break it up.

OK, I think this is ready for a look.  What I've tested:

 - hand deploy new auth, router-resolver to default,
 - tested login/logout flow on web (auth.hail.is/login, /logout) and hailctl (hailctl auth login/logout)
 - then deploy in my namespace:

```
hailctl dev deploy -b cseed/hail:auth -s deploy_auth,deploy_router,deploy_notebook2
```

 - and test login/logout flow via notebook2 (internal.hail.is/cseed/notebook2, etc.) and hailctl, where access to internal is mediated by production (default namespace) credentials.  Note, to do this I copied the production oauth2 key to my namespace.  We shouldn't do this in general and should create a shared dev oauth2 key.  Alternatively, we should create a separate login flow doesn't use oauth2 but uses production credentials.
 - and interactively tested notebook2 creating notebooks (but haven't tested the config of the notebooks themselves).

Summary of changes:
 - auth service that handles login/logout flow via Google OAuth2 and user verification via /userdata endpoint.  Web sessions are stored in the aiohttp_session cookie (encrypted), command line sessions are stored in tokens file: tokens.json.  Token files potentially contain tokens for multiple namespaces (e.g. default and cseed in the example workflow above).
 - sessions are now started in the database, table `users.sessions`, which have session_id (32 random bytes, base64-encoded), user_id, creation time and max_age (for expiry)
 - I write notebook2 to use our async stack
 - added a notion of "deploy config" that has three parts: location (one of external, k8s or gce), default_namespace (the default namespace to find services), and service_namespace (of overrides for specific services ... so e.g. you can use the default auth with batch in cseed).  deploy_config main function is to construct URLs to contact services.
 - JWTs and the jwt secret key are gone.
 - Simplified configuration/data file handling by enforcing consistent defaults.  File paths should be determined by the location, which is loaded from HAIL_DEPLOY_CONFIG_FILE.  If that isn't set, I look in ~/.hail/deploy_config.json, and if that doesn't exist, use external/default.  All other configuration files are determined by the location: the tokens file is in ~/.hail/tokens.json for external, in /user-tokens/tokens.json for k8s, etc.

What remains:
 - what a `hailctl dev config` to set the (local) deploy config for switching between default and dev namespaces.
 - salt session IDs in the database
 - dev oauth2 key
 - add `dev deploy` service override option so we can use production/default auth with services deployed in dev namespaces.

These are all pretty easy.

akotlar commented on auth/auth/auth.py:
```diff
@@ -0,0 +1,233 @@
+    dbpool = request.app['dbpool']
+    async with dbpool.acquire() as conn:
+        async with conn.cursor() as cursor:
+            await cursor.execute('SELECT * from user_data where user_id = %s;', f'google-oauth2|{id}')
```
This is right, but I want to make sure that we don't want to provision for multiple login methods tied to a single account. If we did want that it may be better not to have a single user_id, but google_id, etc. Downside there is schema change for a new connection, but that doesn't seem like a big performance issue, unless we have a billion users. Or, we can go schema less  route, and use JSON_EXTRACT, but that seems more complicated. If first route (multiple user_id's), this is down to a naming issue (google_user_id).

cseed replied:
I'm happy to just support Google now, and have one Google account per Hail account.

I would say we shouldn't try to envision every potential use case now, but rather make this code easy to change in the future (e.g. by making database migrations easy, something on the todo list).

akotlar replied:
Additionally, I think we should implement a refresh mechanism, maybe as a step 2. Every time the user visits the site with a valid session, we should mark their last visit date, and adjust the session expiration date based on that. This could also allow us to use much shorter sessions. To help prevent a malicious use of this, could set a flag on the session, that states whether it can be rehydrated. In particular, if I visit the site with a valid session (meaning I see my username/profile), I expect that the next action I take is also valid. Our current mechanism will have users that see a perfectly valid page one second, and then get a login redirect the next.

akotlar commented on auth/auth/auth.py:
```diff
@@ -0,0 +1,233 @@
+    session_id = userdata['session_id']
+    async with dbpool.acquire() as conn:
+        async with conn.cursor() as cursor:
+            await cursor.execute('DELETE FROM sessions WHERE session_id = %s;', session_id)
```
We may want to do something besides deleting sessions.

I may be interested in knowing how many sessions a user has had, distribution of session creation times, approximate geographic locations.

What do you think about setting a boolean (valid) flag on the session, or just relying on logout (more normalized): (login: Date, logout: Date, ip: ).

Then on login we can control whether we want to grant the user a session, or alert them to some malarkey.

cseed replied:
Yes, great idea.  I think there are ultimately going to be compliance requirements that track logs (as well as other things, like access) so I think we should revisit this question when those requirements are more clear.

akotlar commented on auth/auth/auth.py:
```diff
@@ -0,0 +1,233 @@
+async def rest_login(request):
+    callback_port = request.query['callback_port']
+
+    flow = get_flow(f'http://127.0.0.1:{callback_port}/oauth2callback')
```
In this scenario it looks like the user is running auth.py locally? If so, that may not be something we allow to happen at all.

Also, is the callback always going to be localhost?

akotlar replied:
And should this be `http://127.0.0.1:{callback_port}/api/v1alpha/oauth2callback`?

cseed replied:
> running auth.py locally 

Which auth.py do you mean here?  It's not running auth/auth.py service, the login command runs its own server to accept the callback.  FYI, I double-checked and gcloud uses an http (non-secure) callback:

```
$ gcloud auth login
Your browser has been opened to visit:

    https://accounts.google.com/o/oauth2/auth?redirect_uri=http%3A%2F%2Flocalhost%3A8085%2F&...
```

No, it's calling the server in login.py:

```
@routes.get('/oauth2callback')
async def callback(request):
    ...
```

Since those two pieces of code are coupled, I don't think versioning is necessary.

akotlar replied:
I've read up on this a bit more, and it seems that the way google protects against this depends on the presence of a client secret.

The solution for clients that don't have the secret is something called PKCE, in which a dynamic secret is created by the client (here the requestor), with each request for an authorization code.

The secret is a long random number, which is then hashed by SHA256. This is sent with the authorization url request. The authorization server logs this hashed secret and returns a valid url. The client then uses this url?code_verifier=sha256(secret). The token-generating server, looks up the hashed secret stored for this authorization url, and verifies that the codes match.

This way, if an attacker is able to set itself as the callback for authorization_url, it cannot use that authorization_url to grab a valid token.

This attack requires a long list of preconditions, which may seem to us ridiculous, but I feel that if a standard method exists to protect against this, we should use it.

https://tools.ietf.org/html/rfc7636
 +~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~+
    | End Device (e.g., Smartphone)  |
    |                                |
    | +-------------+   +----------+ | (6) Access Token  +----------+
    | |Legitimate   |   | Malicious|<--------------------|          |
    | |OAuth 2.0 App|   | App      |-------------------->|          |
    | +-------------+   +----------+ | (5) Authorization |          |
    |        |    ^          ^       |        Grant      |          |
    |        |     \         |       |                   |          |
    |        |      \   (4)  |       |                   |          |
    |    (1) |       \  Authz|       |                   |          |
    |   Authz|        \ Code |       |                   |  Authz   |
    | Request|         \     |       |                   |  Server  |
    |        |          \    |       |                   |          |
    |        |           \   |       |                   |          |
    |        v            \  |       |                   |          |
    | +----------------------------+ |                   |          |
    | |                            | | (3) Authz Code    |          |
    | |     Operating System/      |<--------------------|          |
    | |         Browser            |-------------------->|          |
    | |                            | | (2) Authz Request |          |
    | +----------------------------+ |                   +----------+
    +~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~+

This may be a little weird to use because we construct a single instance of Flow (we would need one per api/ v1alpha/login request).

I thought we could use an inverted process, or if we solve the above, as an additional layer of protection: we use a dynamic RSA encryption. Every time the cli client calls hailtcl auth login, the login.py server creates a public/private key pair. It sends the public key along with the request for the authorization_url (to api server), the api server encrypts the response using this public key, and only the possessor of the private key can decrypt that response. Seems safer.

akotlar replied:
In this vein, apparently item recommends PKCE for all clients, native or not: 

https://tools.ietf.org/html/draft-ietf-oauth-security-topics-11#section-2.1.1
```
  Note: although PKCE so far was recommended as a mechanism to protect
   native apps, this advice applies to all kinds of OAuth clients,
   including web applications.
```

cseed replied:
I'm having trouble correlating your references with the Google flow code.  Here is my best guess: I think we have a client secret (which Google calls state).  Google's `Flow.authorization_url` returns the authorization url along with a state.  I think state is a random token.  Here is what the Google docs say about the state:

    The state is used when completing the flow to verify that the request originated from your application.

https://google-auth-oauthlib.readthedocs.io/en/latest/reference/google_auth_oauthlib.flow.html#google_auth_oauthlib.flow.Flow.authorization_url

The state is returned by auth/api/v1alpha/login to the client (encrypted).  Then the user does the login flow with Google using the authorization url but NOT the state, and the client gets an (local, unencrypted) callback from the browser with the "code".  The client then makes an (encrypted) callback to auth/api/v1alpha/oauth2callback with the state (which was never exposed) and the code.  /oauth2callback then completes the verification with Google using both the state and the code by calling `Flow.get_token` to get the final JWT.  Attacker who captures the code cannot complete the oauth2 flow without the state.  I have verified this (by not propagating the state in /oauth2callback.)

 > This may be a little weird to use because we construct a single instance of Flow (we would need one per api/ v1alpha/login request).

No, we create a Flow per call to auth/api/v1alpha/login by calling `get_flow`.

akotlar replied:
> No, we create a Flow per call to auth/api/v1alpha/login by calling get_flow.

Oops, of course. So we could use this PKCE approach.

code_verifier = kwargs.pop('code_verifier', None)
        autogenerate_code_verifier = kwargs.pop(
            'autogenerate_code_verifier', None)

> The client then makes an (encrypted) callback to auth/api/v1alpha/oauth2callback with the state (which was never exposed) and the code. /oauth2callback then completes the verification with Google using both the state and the code by calling Flow.get_token to get the final JWT

Wait, on the cli side, no call to /auth/api/v1alpha/oauth2callback is made right? login.py gets a response with the authorization url, which is set to /oauth2callback on the cli's local server.

> The state is returned by auth/api/v1alpha/login to the client (encrypted).

From my understanding the PKCE implementation is orthogonal. 

The state secret generated by [requests_outhlib (used by google flow)](https://requests-oauthlib.readthedocs.io/en/stable/api.html#requests_oauthlib.OAuth2Session.authorization_url) is used for CSRF protection (i.e an attacker who could hijack a session, but not read the contents of the cookie that contains the state because it's httpOnly couldn't use the authorization_url).

PKCE is like inverse CSRF. The client creates the secret, and never send it out (except encrypted and over https), so even if the authorization_url is intercepted, it is useless.

In the google flow code, we distinguish between state and `code_verifier`:

```python
def authorization_url(self, **kwargs):
        """Generates an authorization URL.

        This is the first step in the OAuth 2.0 Authorization Flow. The user's
        browser should be redirected to the returned URL.

        This method calls
        :meth:`requests_oauthlib.OAuth2Session.authorization_url`
        and specifies the client configuration's authorization URI (usually
        Google's authorization server) and specifies that "offline" access is
        desired. This is required in order to obtain a refresh token.

        Args:
            kwargs: Additional arguments passed through to
                :meth:`requests_oauthlib.OAuth2Session.authorization_url`

        Returns:
            Tuple[str, str]: The generated authorization URL and state. The
                user must visit the URL to complete the flow. The state is used
                when completing the flow to verify that the request originated
                from your application. If your application is using a different
                :class:`Flow` instance to obtain the token, you will need to
                specify the ``state`` when constructing the :class:`Flow`.
        """
        kwargs.setdefault('access_type', 'offline')
        if self.autogenerate_code_verifier:
            chars = ascii_letters+digits+'-._~'
            rnd = SystemRandom()
            random_verifier = [rnd.choice(chars) for _ in range(0, 128)]
            self.code_verifier = ''.join(random_verifier)

        if self.code_verifier:
            code_hash = hashlib.sha256()
            code_hash.update(str.encode(self.code_verifier))
            unencoded_challenge = code_hash.digest()
            b64_challenge = urlsafe_b64encode(unencoded_challenge)
            code_challenge = b64_challenge.decode().split('=')[0]
            kwargs.setdefault('code_challenge', code_challenge)
            kwargs.setdefault('code_challenge_method', 'S256')
        url, state = self.oauth2session.authorization_url(
            self.client_config['auth_uri'], **kwargs)

        return url, state
```

akotlar replied:
Tracked use; so from here, google calls https://github.com/requests/requests-oauthlib/blob/master/requests_oauthlib/oauth2_session.py#L154, which passes through the code_challenge and code_challenge_method to https://github.com/oauthlib/oauthlib, which uses it in their PKCE implementation (and separately from the CSRF stuff).

akotlar commented:
Cotton, mostly looks great, I haven't taken a look at the deployment configs yet. I would like to do that, and if you're ok with this from a time standpoint, spin up a locally deployed version to play with. 

Nice work!

akotlar commented on batch/Makefile:
```diff
@@ -28,7 +28,7 @@ install:
 
 flake8-stmp: $(PY_FILES)
-	$(PYTHON) -m flake8 batch
+	$(PYTHON) -m flake8  --config ../setup.cfg batch
```
On lines 90 and 100, we still have HAIL_TOKEN_FILE

akotlar commented:
HAIL_TOKEN_FILE still exists in 

```sh
apiserver/test-apiserver-pod.yaml
21:     - name: HAIL_TOKEN_FILE

notebook2/notebook/notebook.py
73:                env=[kube.client.V1EnvVar(name='HAIL_TOKEN_FILE',

notebook2/notebook/kubeclient.py
46:                env=[kube.client.V1EnvVar(name='HAIL_TOKEN_FILE',

hail/python/hail/backend/backend.py
216:                          os.environ.get('HAIL_TOKEN_FILE') or
```

akotlar commented on hail/python/hailtop/hailctl/auth/login.py:
```diff
@@ -0,0 +1,99 @@
+from hailtop.gear.auth import get_tokens, auth_headers
+
+
+def init_parser(parser):  # pylint: disable=unused-argument
```
Could we add tests for login, logout?

cseed replied:
I don't actually know how to do that, it would involve automating the Google oauth2 flow.  I'm not actually sure how to handle it.  I'm sure it's possible, so one option is to automate.  One option is to make this a manual test that someone has to run in a test namespace and attest to it.  Another option is a staging-like step as part of deploy where things like this and testing the UI more generally gate production deployment.  I think the second option is good for now, and we can continue to think about it.

akotlar replied:
Sure, #2 sounds reasonable. I think the typical way to do this would be to mock the service. Google uses this approach: https://github.com/googleapis/google-auth-library-python-oauthlib/blob/master/tests/test_flow.py


```python
authorization_url_spy.assert_called_with(
                CLIENT_SECRETS_INFO['web']['auth_uri'],
                access_type='offline',
                prompt='consent')
```

akotlar commented on hail/python/hailtop/gear/auth/utils.py:
```diff
@@ -0,0 +1,29 @@
+import base64
+
+
+async def create_user(dbpool, spec):
```
So this doesn't actually create a user, it inserts one. This also seems to overlap with my [create user PR](https://github.com/hail-is/hail/pull/6619). How would you like to handle that?

akotlar commented on auth/create-auth-tables.sql:
```diff
@@ -0,0 +1,29 @@
+  KEY `username` (`username`)
+) ENGINE=InnoDB;
+
+CREATE TABLE `sessions` (
```
I think this could potentially be improved with a trigger that automatically deletes itself on expiration (or in the future, takes another action, like setting an expired field). I may not want to let a future developer accidentally provide an expired session to a user (say by forgetting to check max_age).

cseed replied:
I couldn't see how to create a a trigger here, but I added an event every 5 minutes to delete expired sessions.

akotlar replied:
Looks great!

akotlar commented on hail/python/hailtop/gear/deploy_config.py:
```diff
@@ -0,0 +1,98 @@
+log = logging.getLogger('gear')
+
+
+class DeployConfig:
```
I think we should have tests for this.

cseed replied:
I think this is well tested because deploy_config is used by both the clients and the servers.  I added tests for the case of default_namespace=default since that's not actually exercised in the tests (and arguably the most important).

akotlar commented on hail/python/hailtop/hailctl/auth/logout.py:
```diff
@@ -0,0 +1,36 @@
+from hailtop.gear.auth import get_tokens, auth_headers
+
+
+def init_parser(parser):  # pylint: disable=unused-argument
```
Similarly, I think we should have tests for this.

akotlar commented on notebook2/notebook/notebook.py:
```diff
@@ -301,283 +175,210 @@ def pod_condition_for_ui(conds):
-def worker_image():
-    return WORKER_IMAGE, 200
+@routes.get('')
+@routes.get('/')
```
duplicate

cseed replied:
This is not duplicate.  It's a sort of workaround for aiohttp.  We're using subapps when running on in a dev namespace (to add the /ns/svc path prefix), and if you navigate to internal.hail.is/ns/svc, the path is '' in the subapp and it doesn't match '/'.

akotlar replied:
oh weird

akotlar commented on notebook2/notebook/notebook.py:
```diff
@@ -301,283 +175,210 @@ def pod_condition_for_ui(conds):
+@routes.get('')
+@routes.get('/')
+@aiohttp_jinja2.template('index.html')
+@web_maybe_authenticated_user
```
Currently on each route, we do a database lookup, not only for the session, but for the user's data. This feels like a potentially significant performance regression. I would prefer that upon login the user's data gets cached somehow, probably in the browser (via localStorage or httpOnly cookie) for now.

cseed replied:
We do one database query that involves an indexed join.  The overhead for the join is ~150us:

```
0.00038250 | select * from sessions where session_id = <session>
0.00054050 | select * from sessions inner join user_data on user_data.id = sessions.user_id where session_id = <session>
```

I'm not inclined to spend time optimizing this.

akotlar replied:
Thanks for the performance data

akotlar commented on auth/deployment.yaml:
```diff
@@ -0,0 +1,110 @@
+apiVersion: apps/v1beta2
```
I think we should have component deployment tests. Could be part of a larger integrated test.

cseed replied:
What's a component deployment test?

akotlar replied:
Test that the component/service can be deployed successfully to Kubernetes

cseed replied:
I'm still not sure I understand.  All the components are deployed during the tests.

akotlar replied:
I thought auth wasn't being tested (the integrated test would be what you described), I think because I was expecting it to be tested directly. So batch is testing auth endpoints, and we specify the dependency to deploy auth in the testing namespace. 

I do see a deploy_auth dependency in deploy_batch, which makes deploy_auth a transitive dependency of

```sh
 - kind: deploy
   name: test_batch
   namespace:
     valueFrom: batch_pods_ns.name
   config: batch/test-batch-pod.yaml
   wait:
    - kind: Pod
      name: test-batch
      for: completed
      timeout: 900
   dependsOn:
    - default_ns
    - batch_pods_ns
    - deploy_batch
    - deploy_batch_pods
    - test_batch_image
```

That wasn't immediately obvious to me, sorry.

cseed replied:
> I thought auth wasn't being tested

No problem, happy to clarify.  Auth is being partially tested.  Auth is now a lynchpin and nothing can function without it.  Auth is being deployed in tests and used by all the clients and services for authorization.  The login/logout flow is not being tested.  The way we get accounts without login is by creating sessions directly in the create_accounts step and saving them to <user>-tokens secrets (like we did with JWTs before).

akotlar commented on auth/auth/auth.py:
```diff
@@ -0,0 +1,233 @@
+            await cursor.execute('''
+SELECT user_data.*, sessions.session_id FROM user_data
+INNER JOIN sessions ON user_data.id = sessions.user_id
+WHERE (sessions.session_id = %s) AND (ISNULL(sessions.max_age_secs) OR (NOW() < TIMESTAMPADD(SECOND, sessions.max_age_secs, sessions.created)));
```
max_age_secs should never be null right? This logic also means that logout and session check use somewhat different semantics (logout deletes, but we'll have a bunch of dangling sessions because a user whose session is invalidated is effectively logged out, without actually calling the logout endpoint).

I think we could solve this by string an expired field on the session entry. (or I suppose deleting the session here).

akotlar replied:
Yes, so I think the 2nd way to solve this dangling session issue is to changed _authenticated_users_only to always delete expired sessions. This could be done by forcing a redirect to /logout, but then we'd have dangling sessions with `@web_maybe_authenticated_users`

cseed replied:
> max_age_secs should never be null right?

Until we have key rotation in place, I was planning for the user user session secrets in K8s to have no expiration, hence explicitly supporting NULL max_age_secs.

akotlar commented on auth/create-auth-tables.sql:
```diff
@@ -0,0 +1,29 @@
+  `session_id` VARCHAR(255) NOT NULL,
+  `user_id` INT(11) NOT NULL,
+  `max_age_secs` INT(11) UNSIGNED DEFAULT NULL,
+  `created` TIMESTAMP DEFAULT NOW(),
```
Why does session_id need to be VARCHAR? May have more trouble indexing it, and wouldn't a cryptographically secure random number be sufficient?

cseed replied:
It's a 256-bit random string (which I base64 encode).  I don't see how to turn that into a number.  Also, MySQL indexes VARCHAR, so I don't see a problem here.

akotlar commented:
Im also happy to let this in as mostly is (so no additional tests, api changes), and set the remainder that you find valid as issues. No need to stall something that isn't very user facing.

cseed commented:
OK, I think I addressed all the comments.  Here is a summary of the code changes I made:
 - use *-tokens instead of *-jwt for the session tokens
 - db event to clean up sessions,
 - there was a bunch of legacy garbage in batch/ and batch/Makefile that I nuked,
 - added hailtop test for deploy_config,
 - fixed up hail ServiceBackend (which isn't actually tested now)
 - create_user => insert_user

I think there is still some legacy garbage in apiserver/, but that's not being deployed right now so I just left it.

Let me know if I missed anything.
